### ISSUE TITLE
Example of using the standalone application will be great

### ISSUE THREAD

**Issue by ameetchhatwal** [fullstorydev/grpcui#53]
It will be great to have an example of how to use the standalone application.

**Comment by jhump** [comment]
@ameetchhatwal, can you elaborate what you mean? The README already has examples of using it. Did you instead mean Go code examples of using the standalone _package_?

**Comment by ameetchhatwal** [comment]
Hi @jhump Thanks for responding back, I already have a grpc service. I was able to use the cmd line grpcui tool to access my service, but I wanted to explore if I can natively deploy grpcui with the service so I can just access grpcui on a different port. 
From looking at the Readme standalone app seems like a possibility, but any example or guide of how to integrate it will be helpful. Please let me know if you any follow up question.

**Comment by jhump** [comment]
@ameetchhatwal, I still don't understand. The README makes no mention of "standalone app". It has a single reference to the standalone Go package though. Are you asking how to embed the gRPC UI _into_ your Go gRPC server? (That's what that package is for.)

If not, and you just want to deploy grpcui _with_ your server, you can easily do so: just have them both running on different ports with grpcui configured to talk directly to the co-located server over localhost.

**Comment by ameetchhatwal** [comment]
Yes, I wanna embed grpcui into my go grpc service so when I deploy it, I
can access service as it is today but also have access to the GRPC UI which
talks to my service.
I am not sure how can I do that?

Sorry for the confusion.

**Comment by jhump** [comment]
There is an example in the Go docs for that package:
https://godoc.org/github.com/fullstorydev/grpcui/standalone

In that example, you could have it dial "localhost" to connect to the server that is running in the same process. Also, for that code sample to work, you'll need to enable service reflection in your server: https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#enable-server-reflection

The "Cadillac" integration would use an in-process channel, avoiding the need to connect over loopback to the gRPC server. And it would also not require service reflection, since it's wasteful for the client to download and de-serialize descriptors from the server when they are in the same process -- the client should just share the descriptors linked into the program. Going this path is much less on rails. Here's what it entails, in fairly broad strokes:
1. Use `github.com/fullstorydev/grpchan` and the included `protoc-gen-grpchan` plugin to generate alternate service registration and stub factory functions from your protos. This allows you to register services in something other than a `*grpc.Server` and to create client stubs around something other than a `*grpc.ClientConn`:
https://godoc.org/github.com/fullstorydev/grpchan
2. When you register your services in your actual gRPC server, also register them with an in-process channel: https://godoc.org/github.com/fullstorydev/grpchan/inprocgrpc#Channel
3. Use `grpcui.AllMethodsForServer` to extract method descriptors for everything registered in your gRPC server:
https://godoc.org/github.com/fullstorydev/grpcui#AllMethodsForServer
4. The most complicated part: accumulate all file descriptors for the methods retrieved from the previous step. I'll include a code example of how one could do this below. Admittedly, this would probably be useful as a helper function in the `grpcui` package.
5. Now you can create a standalone handler using `standalone.Handler` (_not_ `HandlerViaReflection`). Pass the in-process channel as well as the method and file descriptors retrieved in the steps above.
https://godoc.org/github.com/fullstorydev/grpcui/standalone#Handler

The least "on rails" step is bullet 4. This is basically what that code needs to look like:

```go
// GetAllFiles accepts a slice of method descriptors and returns a slice of
// file descriptors representing the transitive dependencies of all files in
// which the given methods are defined.
func GetAllFiles(methods []*desc.MethodDescriptor) []*desc.FileDescriptor {
	fileSet := map[*desc.FileDescriptor]struct{}{}
	for _, md := range methods {
		addToFiles(md.GetFile(), fileSet)
	}
	files := make([]*desc.FileDescriptor, 0, len(fileSet))
	for fd := range fileSet {
		files = append(files, fd)
	}
	return files
}

func addToFiles(f *desc.FileDescriptor, files map[*desc.FileDescriptor]struct{}) {
	if _, ok := files[f]; ok {
		// already added
		return
	}
	files[f] = struct{}{}
	for _, dep := range f.GetDependencies() {
		addToFiles(dep, files)
	}
}
```

**Comment by jhump** [comment]
@ameetchhatwal, if the above is what you need, feel free to close this issue.
